### ISSUE TITLE
Fix enums and typedefs in `icrate`

### ISSUE THREAD

**Issue by madsmtm** [madsmtm/objc2#310]
Right now we're just using type aliases - instead, we should generate newtypes with the appropriate traits and constants.

**Comment by silvanshade** [comment]
I believe the implementation I've been working on in #425 does a reasonably good job now of handling enums at least.

I generally tried to follow a similar approach for mapping things over to how the enums are translated for Swift. There are some caveats, for example due to the fact that we can't define an `enum` within a `struct` in Rust like you can in Swift, or not having `impl` for `enum`, or not (yet) having associated types for inherent impls.

The most sophisticated example is `icrate/examples/ns_error_enum.rs` where I use the macro to create an error enum translation for `NSURLError`. (See the code snippet and macro expansion at the end for reference).

A couple things to note:

You can (optionally) specify the name of the helper module with `mod`. I think it might be reasonable to update the macro with a list of Apple system framework prefixes and special cases so this isn't needed (the only reason it's used here is for correct tokenization of `NSURL`). 

You can also (optionally) specify refinements for the `userInfo` dictionary for the error. The syntax is `{ key = <ident>, type = <type>, from? = <expr>, into? = <expr> }`. The `key` is expected to be an identifier in scope that resolves to a dictionary key. The `type` is the type of the associated field in the generated `UserInfo` struct (which is then wrapped in `Option<...>`). The optional `from` and `into` keys are used to convert the entry _from_ the `NSDictionary` to the `UserInfo` field, and then back from the field _into_ the entry.

Similar to Swift, getters are generated for the specified `UserInfo` fields, but you can also pattern match on the entire `UserInfo` struct (which uses the getters to populate) like so:

```rust
// `UserInfo<T>` is an alias for `<T as objc2::ErrorEnum>::UserInfo` (to avoid mentioning the helper module)
let UserInfo::<NSURLError> {
    failing_url,
    failing_url_string,
    background_task_cancelled_reason,
    network_unavailable_reason,
} = error.user_info();
let failing_url_string = error.failing_url_string();
```

The error declaration looks like this:

```rust
#[objc(error_enum,
    mod = ns_url_error,
    domain = unsafe { NSURLErrorDomain },
    user_info = [
        // NOTE: Here we specify (typed) getters for the associated `userInfo` dict.
        { key = NSURLErrorFailingURLErrorKey, type = Id<NSURL> },
        { key = NSURLErrorFailingURLStringErrorKey, type = Id<NSString> },
        { key = NSURLErrorBackgroundTaskCancelledReasonKey, type = ns_url_error_reasons::BackgroundTaskCancelledReason },
        { key = NSURLErrorNetworkUnavailableReasonKey, type = ns_url_error_reasons::NetworkUnavailableReason },
    ]
)]
pub enum NSURLError {
    Unknown = -1,
    Cancelled = -999,
    BadURL = -1000,
    TimedOut = -1001,
    UnsupportedURL = -1002,
    CannotFindHost = -1003,
    CannotConnectToHost = -1004,
    NetworkConnectionLost = -1005,
    DNSLookupFailed = -1006,
    HTTPTooManyRedirects = -1007,
    ResourceUnavailable = -1008,
    NotConnectedToInternet = -1009,
    RedirectToNonExistentLocation = -1010,
    BadServerResponse = -1011,
    UserCancelledAuthentication = -1012,
    UserAuthenticationRequired = -1013,
    ZeroByteResource = -1014,
    CannotDecodeRawData = -1015,
    CannotDecodeContentData = -1016,
    CannotParseResponse = -1017,
    AppTransportSecurityRequiresSecureConnection = -1022,
    FileDoesNotExist = -1100,
    FileIsDirectory = -1101,
    NoPermissionsToReadFile = -1102,
    DataLengthExceedsMaximum = -1103,
    FileOutsideSafeArea = -1104,
    SecureConnectionFailed = -1200,
    ServerCertificateHasBadDate = -1201,
    ServerCertificateUntrusted = -1202,
    ServerCertificateHasUnknownRoot = -1203,
    ServerCertificateNotYetValid = -1204,
    ClientCertificateRejected = -1205,
    ClientCertificateRequired = -1206,
    CannotLoadFromNetwork = -2000,
    CannotCreateFile = -3000,
    CannotOpenFile = -3001,
    CannotCloseFile = -3002,
    CannotWriteToFile = -3003,
    CannotRemoveFile = -3004,
    CannotMoveFile = -3005,
    DownloadDecodingFailedMidStream = -3006,
    DownloadDecodingFailedToComplete = -3007,
    InternationalRoamingOff = -1018,
    CallIsActive = -1019,
    DataNotAllowed = -1020,
    RequestBodyStreamExhausted = -1021,
    BackgroundSessionRequiresSharedContainer = -995,
    BackgroundSessionInUseByAnotherProcess = -996,
    BackgroundSessionWasDisconnected = -997,
}
```

which expands to the following:

```rust
// Recursive expansion of objc! macro
// ===================================

#[cfg(all(
    feature = "Foundation",
    feature = "Foundation_NSError",
    feature = "Foundation_NSString"
))]
objc2::declare_class!(
    pub struct NSURLError {}
    unsafe impl ClassType for NSURLError {
        type Super = icrate::Foundation::NSError;
        const NAME: &'static str = "NSURLError";
    }
);
#[cfg(all(
    feature = "Foundation",
    feature = "Foundation_NSError",
    feature = "Foundation_NSString"
))]
impl NSURLError {
    #[inline]
    pub fn domain() -> &'static icrate::Foundation::NSErrorDomain {
        unsafe { NSURLErrorDomain }
    }
    #[inline]
    pub fn code(&self) -> ns_url_error::Codes {
        ns_url_error::Codes(self.as_super().code())
    }
    #[inline]
    pub fn localized_description(&self) -> objc2::rc::Id<icrate::Foundation::NSString> {
        self.as_super().localizedDescription()
    }
    #[inline]
    pub fn user_info(&self) -> ns_url_error::UserInfo {
        let mut user_info = ns_url_error::UserInfo::default();
        user_info.failing_url = self.failing_url();
        user_info.failing_url_string = self.failing_url_string();
        user_info.background_task_cancelled_reason = self.background_task_cancelled_reason();
        user_info.network_unavailable_reason = self.network_unavailable_reason();
        user_info
    }
    #[inline]
    pub fn new(code: ns_url_error::Codes) -> objc2::rc::Id<Self, objc2::rc::Shared> {
        let code = code.0;
        let domain = Self::domain();
        let error = unsafe { icrate::Foundation::NSError::new(code, domain) };
        unsafe { objc2::rc::Id::cast(error) }
    }
    #[cfg(all(
        feature = "Foundation_NSDictionary",
        feature = "Foundation_NSMutableDictionary"
    ))]
    #[inline]
    fn new_with_user_info(
        code: ns_url_error::Codes,
        user_info: ns_url_error::UserInfo,
    ) -> objc2::rc::Id<Self, objc2::rc::Shared> {
        let domain = Self::domain();
        let code = code.0;
        let dict = Self::user_info_into_dictionary(user_info);
        let error = unsafe {
            icrate::Foundation::NSError::errorWithDomain_code_userInfo(domain, code, Some(&*dict))
        };
        unsafe { objc2::rc::Id::cast(error) }
    }
    #[doc(hidden)]
    #[inline]
    fn user_info_into_dictionary(
        user_info: ns_url_error::UserInfo,
    ) -> objc2::rc::Id<
        icrate::Foundation::NSMutableDictionary<
            icrate::Foundation::NSErrorUserInfoKey,
            objc2::runtime::Object,
        >,
    > {
        let dict = icrate::Foundation::NSMutableDictionary::<
            icrate::Foundation::NSErrorUserInfoKey,
            objc2::runtime::Object,
        >::new();
        if let Some(value) = (user_info.failing_url) {
            unsafe { dict.setValue_forKey(Some(&*value), NSURLErrorFailingURLErrorKey) };
        }
        if let Some(value) = (user_info.failing_url_string) {
            unsafe { dict.setValue_forKey(Some(&*value), NSURLErrorFailingURLStringErrorKey) };
        }
        if let Some(value) = (user_info.background_task_cancelled_reason) {
            unsafe {
                dict.setValue_forKey(Some(&*value), NSURLErrorBackgroundTaskCancelledReasonKey)
            };
        }
        if let Some(value) = (user_info.network_unavailable_reason) {
            unsafe { dict.setValue_forKey(Some(&*value), NSURLErrorNetworkUnavailableReasonKey) };
        }
        objc2::rc::Id::into_shared(dict)
    }
    #[inline]
    fn failing_url(&self) -> Option<Id<NSURL>> {
        let user_info = self.as_super().userInfo();
        let value = unsafe { user_info.valueForKey(NSURLErrorFailingURLErrorKey) }
            .map(|inner| unsafe { objc2::rc::Id::cast(inner) });
        (|n: Option<_>| n.map(TryInto::try_into).and_then(Result::ok))(value)
    }
    #[inline]
    fn failing_url_string(&self) -> Option<Id<NSString>> {
        let user_info = self.as_super().userInfo();
        let value = unsafe { user_info.valueForKey(NSURLErrorFailingURLStringErrorKey) }
            .map(|inner| unsafe { objc2::rc::Id::cast(inner) });
        (|n: Option<_>| n.map(TryInto::try_into).and_then(Result::ok))(value)
    }
    #[inline]
    fn background_task_cancelled_reason(
        &self,
    ) -> Option<ns_url_error_reasons::BackgroundTaskCancelledReason> {
        let user_info = self.as_super().userInfo();
        let value = unsafe { user_info.valueForKey(NSURLErrorBackgroundTaskCancelledReasonKey) }
            .map(|inner| unsafe { objc2::rc::Id::cast(inner) });
        (|n: Option<_>| n.map(TryInto::try_into).and_then(Result::ok))(value)
    }
    #[inline]
    fn network_unavailable_reason(&self) -> Option<ns_url_error_reasons::NetworkUnavailableReason> {
        let user_info = self.as_super().userInfo();
        let value = unsafe { user_info.valueForKey(NSURLErrorNetworkUnavailableReasonKey) }
            .map(|inner| unsafe { objc2::rc::Id::cast(inner) });
        (|n: Option<_>| n.map(TryInto::try_into).and_then(Result::ok))(value)
    }
}
#[cfg(all(
    feature = "Foundation",
    feature = "Foundation_NSError",
    feature = "Foundation_NSString"
))]
impl objc2::ErrorEnum for NSURLError {
    type Codes = ns_url_error::Codes;
    type UserInfo = ns_url_error::UserInfo;
}
#[cfg(all(
    feature = "Foundation",
    feature = "Foundation_NSError",
    feature = "Foundation_NSString"
))]
impl objc2::TypedEnum for NSURLError {
    type Cases = ns_url_error::Cases;
}
#[allow(non_snake_case)]
pub mod ns_url_error {
    #[derive(Eq, PartialEq)]
    #[repr(transparent)]
    pub struct Codes(pub(super) isize);

    impl Codes {
        #![allow(non_upper_case_globals)]
        pub const Unknown: Self = Self(-1);
        pub const Cancelled: Self = Self(-999);
        pub const BadURL: Self = Self(-1000);
        pub const TimedOut: Self = Self(-1001);
        pub const UnsupportedURL: Self = Self(-1002);
        pub const CannotFindHost: Self = Self(-1003);
        pub const CannotConnectToHost: Self = Self(-1004);
        pub const NetworkConnectionLost: Self = Self(-1005);
        pub const DNSLookupFailed: Self = Self(-1006);
        pub const HTTPTooManyRedirects: Self = Self(-1007);
        pub const ResourceUnavailable: Self = Self(-1008);
        pub const NotConnectedToInternet: Self = Self(-1009);
        pub const RedirectToNonExistentLocation: Self = Self(-1010);
        pub const BadServerResponse: Self = Self(-1011);
        pub const UserCancelledAuthentication: Self = Self(-1012);
        pub const UserAuthenticationRequired: Self = Self(-1013);
        pub const ZeroByteResource: Self = Self(-1014);
        pub const CannotDecodeRawData: Self = Self(-1015);
        pub const CannotDecodeContentData: Self = Self(-1016);
        pub const CannotParseResponse: Self = Self(-1017);
        pub const AppTransportSecurityRequiresSecureConnection: Self = Self(-1022);
        pub const FileDoesNotExist: Self = Self(-1100);
        pub const FileIsDirectory: Self = Self(-1101);
        pub const NoPermissionsToReadFile: Self = Self(-1102);
        pub const DataLengthExceedsMaximum: Self = Self(-1103);
        pub const FileOutsideSafeArea: Self = Self(-1104);
        pub const SecureConnectionFailed: Self = Self(-1200);
        pub const ServerCertificateHasBadDate: Self = Self(-1201);
        pub const ServerCertificateUntrusted: Self = Self(-1202);
        pub const ServerCertificateHasUnknownRoot: Self = Self(-1203);
        pub const ServerCertificateNotYetValid: Self = Self(-1204);
        pub const ClientCertificateRejected: Self = Self(-1205);
        pub const ClientCertificateRequired: Self = Self(-1206);
        pub const CannotLoadFromNetwork: Self = Self(-2000);
        pub const CannotCreateFile: Self = Self(-3000);
        pub const CannotOpenFile: Self = Self(-3001);
        pub const CannotCloseFile: Self = Self(-3002);
        pub const CannotWriteToFile: Self = Self(-3003);
        pub const CannotRemoveFile: Self = Self(-3004);
        pub const CannotMoveFile: Self = Self(-3005);
        pub const DownloadDecodingFailedMidStream: Self = Self(-3006);
        pub const DownloadDecodingFailedToComplete: Self = Self(-3007);
        pub const InternationalRoamingOff: Self = Self(-1018);
        pub const CallIsActive: Self = Self(-1019);
        pub const DataNotAllowed: Self = Self(-1020);
        pub const RequestBodyStreamExhausted: Self = Self(-1021);
        pub const BackgroundSessionRequiresSharedContainer: Self = Self(-995);
        pub const BackgroundSessionInUseByAnotherProcess: Self = Self(-996);
        pub const BackgroundSessionWasDisconnected: Self = Self(-997);
        pub fn cases(&self) -> core::option::Option<Cases> {
            match self {
                &Self::Unknown => Cases::Unknown.into(),
                &Self::Cancelled => Cases::Cancelled.into(),
                &Self::BadURL => Cases::BadURL.into(),
                &Self::TimedOut => Cases::TimedOut.into(),
                &Self::UnsupportedURL => Cases::UnsupportedURL.into(),
                &Self::CannotFindHost => Cases::CannotFindHost.into(),
                &Self::CannotConnectToHost => Cases::CannotConnectToHost.into(),
                &Self::NetworkConnectionLost => Cases::NetworkConnectionLost.into(),
                &Self::DNSLookupFailed => Cases::DNSLookupFailed.into(),
                &Self::HTTPTooManyRedirects => Cases::HTTPTooManyRedirects.into(),
                &Self::ResourceUnavailable => Cases::ResourceUnavailable.into(),
                &Self::NotConnectedToInternet => Cases::NotConnectedToInternet.into(),
                &Self::RedirectToNonExistentLocation => Cases::RedirectToNonExistentLocation.into(),
                &Self::BadServerResponse => Cases::BadServerResponse.into(),
                &Self::UserCancelledAuthentication => Cases::UserCancelledAuthentication.into(),
                &Self::UserAuthenticationRequired => Cases::UserAuthenticationRequired.into(),
                &Self::ZeroByteResource => Cases::ZeroByteResource.into(),
                &Self::CannotDecodeRawData => Cases::CannotDecodeRawData.into(),
                &Self::CannotDecodeContentData => Cases::CannotDecodeContentData.into(),
                &Self::CannotParseResponse => Cases::CannotParseResponse.into(),
                &Self::AppTransportSecurityRequiresSecureConnection => {
                    Cases::AppTransportSecurityRequiresSecureConnection.into()
                }
                &Self::FileDoesNotExist => Cases::FileDoesNotExist.into(),
                &Self::FileIsDirectory => Cases::FileIsDirectory.into(),
                &Self::NoPermissionsToReadFile => Cases::NoPermissionsToReadFile.into(),
                &Self::DataLengthExceedsMaximum => Cases::DataLengthExceedsMaximum.into(),
                &Self::FileOutsideSafeArea => Cases::FileOutsideSafeArea.into(),
                &Self::SecureConnectionFailed => Cases::SecureConnectionFailed.into(),
                &Self::ServerCertificateHasBadDate => Cases::ServerCertificateHasBadDate.into(),
                &Self::ServerCertificateUntrusted => Cases::ServerCertificateUntrusted.into(),
                &Self::ServerCertificateHasUnknownRoot => {
                    Cases::ServerCertificateHasUnknownRoot.into()
                }
                &Self::ServerCertificateNotYetValid => Cases::ServerCertificateNotYetValid.into(),
                &Self::ClientCertificateRejected => Cases::ClientCertificateRejected.into(),
                &Self::ClientCertificateRequired => Cases::ClientCertificateRequired.into(),
                &Self::CannotLoadFromNetwork => Cases::CannotLoadFromNetwork.into(),
                &Self::CannotCreateFile => Cases::CannotCreateFile.into(),
                &Self::CannotOpenFile => Cases::CannotOpenFile.into(),
                &Self::CannotCloseFile => Cases::CannotCloseFile.into(),
                &Self::CannotWriteToFile => Cases::CannotWriteToFile.into(),
                &Self::CannotRemoveFile => Cases::CannotRemoveFile.into(),
                &Self::CannotMoveFile => Cases::CannotMoveFile.into(),
                &Self::DownloadDecodingFailedMidStream => {
                    Cases::DownloadDecodingFailedMidStream.into()
                }
                &Self::DownloadDecodingFailedToComplete => {
                    Cases::DownloadDecodingFailedToComplete.into()
                }
                &Self::InternationalRoamingOff => Cases::InternationalRoamingOff.into(),
                &Self::CallIsActive => Cases::CallIsActive.into(),
                &Self::DataNotAllowed => Cases::DataNotAllowed.into(),
                &Self::RequestBodyStreamExhausted => Cases::RequestBodyStreamExhausted.into(),
                &Self::BackgroundSessionRequiresSharedContainer => {
                    Cases::BackgroundSessionRequiresSharedContainer.into()
                }
                &Self::BackgroundSessionInUseByAnotherProcess => {
                    Cases::BackgroundSessionInUseByAnotherProcess.into()
                }
                &Self::BackgroundSessionWasDisconnected => {
                    Cases::BackgroundSessionWasDisconnected.into()
                }
                _ => None,
            }
        }
        #[inline]
        pub fn peek(&self) -> &isize {
            &self.0
        }
        #[inline]
        pub fn take(self) -> isize {
            self.0
        }
    }
    impl From<Cases> for Codes {
        fn from(case: Cases) -> Self {
            match case {
                Cases::Unknown => Codes::Unknown,
                Cases::Cancelled => Codes::Cancelled,
                Cases::BadURL => Codes::BadURL,
                Cases::TimedOut => Codes::TimedOut,
                Cases::UnsupportedURL => Codes::UnsupportedURL,
                Cases::CannotFindHost => Codes::CannotFindHost,
                Cases::CannotConnectToHost => Codes::CannotConnectToHost,
                Cases::NetworkConnectionLost => Codes::NetworkConnectionLost,
                Cases::DNSLookupFailed => Codes::DNSLookupFailed,
                Cases::HTTPTooManyRedirects => Codes::HTTPTooManyRedirects,
                Cases::ResourceUnavailable => Codes::ResourceUnavailable,
                Cases::NotConnectedToInternet => Codes::NotConnectedToInternet,
                Cases::RedirectToNonExistentLocation => Codes::RedirectToNonExistentLocation,
                Cases::BadServerResponse => Codes::BadServerResponse,
                Cases::UserCancelledAuthentication => Codes::UserCancelledAuthentication,
                Cases::UserAuthenticationRequired => Codes::UserAuthenticationRequired,
                Cases::ZeroByteResource => Codes::ZeroByteResource,
                Cases::CannotDecodeRawData => Codes::CannotDecodeRawData,
                Cases::CannotDecodeContentData => Codes::CannotDecodeContentData,
                Cases::CannotParseResponse => Codes::CannotParseResponse,
                Cases::AppTransportSecurityRequiresSecureConnection => {
                    Codes::AppTransportSecurityRequiresSecureConnection
                }
                Cases::FileDoesNotExist => Codes::FileDoesNotExist,
                Cases::FileIsDirectory => Codes::FileIsDirectory,
                Cases::NoPermissionsToReadFile => Codes::NoPermissionsToReadFile,
                Cases::DataLengthExceedsMaximum => Codes::DataLengthExceedsMaximum,
                Cases::FileOutsideSafeArea => Codes::FileOutsideSafeArea,
                Cases::SecureConnectionFailed => Codes::SecureConnectionFailed,
                Cases::ServerCertificateHasBadDate => Codes::ServerCertificateHasBadDate,
                Cases::ServerCertificateUntrusted => Codes::ServerCertificateUntrusted,
                Cases::ServerCertificateHasUnknownRoot => Codes::ServerCertificateHasUnknownRoot,
                Cases::ServerCertificateNotYetValid => Codes::ServerCertificateNotYetValid,
                Cases::ClientCertificateRejected => Codes::ClientCertificateRejected,
                Cases::ClientCertificateRequired => Codes::ClientCertificateRequired,
                Cases::CannotLoadFromNetwork => Codes::CannotLoadFromNetwork,
                Cases::CannotCreateFile => Codes::CannotCreateFile,
                Cases::CannotOpenFile => Codes::CannotOpenFile,
                Cases::CannotCloseFile => Codes::CannotCloseFile,
                Cases::CannotWriteToFile => Codes::CannotWriteToFile,
                Cases::CannotRemoveFile => Codes::CannotRemoveFile,
                Cases::CannotMoveFile => Codes::CannotMoveFile,
                Cases::DownloadDecodingFailedMidStream => Codes::DownloadDecodingFailedMidStream,
                Cases::DownloadDecodingFailedToComplete => Codes::DownloadDecodingFailedToComplete,
                Cases::InternationalRoamingOff => Codes::InternationalRoamingOff,
                Cases::CallIsActive => Codes::CallIsActive,
                Cases::DataNotAllowed => Codes::DataNotAllowed,
                Cases::RequestBodyStreamExhausted => Codes::RequestBodyStreamExhausted,
                Cases::BackgroundSessionRequiresSharedContainer => {
                    Codes::BackgroundSessionRequiresSharedContainer
                }
                Cases::BackgroundSessionInUseByAnotherProcess => {
                    Codes::BackgroundSessionInUseByAnotherProcess
                }
                Cases::BackgroundSessionWasDisconnected => Codes::BackgroundSessionWasDisconnected,
            }
        }
    }
    impl core::convert::From<Codes> for isize {
        #[inline]
        fn from(wrapper: Codes) -> Self {
            wrapper.take()
        }
    }
    impl From<Cases> for isize {
        #[inline]
        fn from(case: Cases) -> Self {
            Self::from(Codes::from(case))
        }
    }
    impl objc2::TypedEnum for Codes {
        type Cases = Cases;
    }
    #[cfg(all(feature = "Foundation", feature = "Foundation_NSNumber"))]
    impl From<Codes> for objc2::rc::Id<icrate::Foundation::NSNumber> {
        #[inline]
        fn from(value: Codes) -> Self {
            icrate::Foundation::NSNumber::new_isize(value.0.into())
        }
    }
    #[cfg(feature = "Foundation")]
    impl core::ops::Deref for Codes {
        type Target = icrate::Foundation::NSObject;
        fn deref(&self) -> &Self::Target {
            let num = icrate::Foundation::NSNumber::new_isize(self.0.into());
            let obj = unsafe { objc2::rc::Id::cast::<icrate::Foundation::NSObject>(num) };
            let obj = objc2::rc::Id::autorelease_return(obj);
            let obj = unsafe { obj.as_ref() }.expect("pointer is non-null");
            obj
        }
    }
    #[cfg(all(feature = "Foundation", feature = "Foundation_NSNumber"))]
    impl TryFrom<&icrate::Foundation::NSNumber> for Codes {
        type Error = Box<dyn std::error::Error + Send + Sync + 'static>;
        fn try_from(number: &icrate::Foundation::NSNumber) -> Result<Self, Self::Error> {
            use objc2::Encoding::*;
            let encoding = number.encoding();
            let equivalent = match "isize" {
                "usize" => {
                    encoding == ULong
                        || (cfg!(target_pointer_width = "32") && encoding == UInt)
                        || (cfg!(target_pointer_width = "64") && encoding == ULongLong)
                }
                "isize" => {
                    encoding == Long
                        || (cfg!(target_pointer_width = "32") && encoding == Int)
                        || (cfg!(target_pointer_width = "64") && encoding == LongLong)
                }
                "u8" => encoding == UChar,
                "u16" => encoding == UShort,
                "u32" => {
                    encoding == UInt || (cfg!(target_pointer_width = "32") && encoding == ULong)
                }
                "u64" => {
                    encoding == ULongLong
                        || (cfg!(target_pointer_width = "64") && encoding == ULong)
                }
                "i8" => encoding == Char,
                "i16" => encoding == Short,
                "i32" => encoding == Int || (cfg!(target_pointer_width = "32") && encoding == Long),
                "i64" => {
                    encoding == LongLong || (cfg!(target_pointer_width = "64") && encoding == Long)
                }
                _ => false,
            };
            if equivalent {
                let number = number.as_isize();
                let value = Self(number);
                if value.cases().is_some() {
                    Ok(value)
                } else {
                    Err(format!(
                        "NSNumber value `{}` does not match any of enum values for `{}`",
                        number, "Codes",
                    )
                    .into())
                }
            } else {
                Err(format!(
                    "NSNumber encoding `{}` is not equivalent to `{}` repr `{}`",
                    encoding, "Codes", "isize",
                )
                .into())
            }
        }
    }
    impl<O: objc2::rc::Ownership> TryFrom<objc2::rc::Id<icrate::Foundation::NSNumber, O>> for Codes {
        type Error = Box<dyn std::error::Error + Send + Sync + 'static>;
        #[inline]
        fn try_from(number: Id<icrate::Foundation::NSNumber, O>) -> Result<Self, Self::Error> {
            (&*number).try_into()
        }
    }
    #[allow(non_camel_case_types)]
    #[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
    #[non_exhaustive]
    pub enum Cases {
        Unknown,
        Cancelled,
        BadURL,
        TimedOut,
        UnsupportedURL,
        CannotFindHost,
        CannotConnectToHost,
        NetworkConnectionLost,
        DNSLookupFailed,
        HTTPTooManyRedirects,
        ResourceUnavailable,
        NotConnectedToInternet,
        RedirectToNonExistentLocation,
        BadServerResponse,
        UserCancelledAuthentication,
        UserAuthenticationRequired,
        ZeroByteResource,
        CannotDecodeRawData,
        CannotDecodeContentData,
        CannotParseResponse,
        AppTransportSecurityRequiresSecureConnection,
        FileDoesNotExist,
        FileIsDirectory,
        NoPermissionsToReadFile,
        DataLengthExceedsMaximum,
        FileOutsideSafeArea,
        SecureConnectionFailed,
        ServerCertificateHasBadDate,
        ServerCertificateUntrusted,
        ServerCertificateHasUnknownRoot,
        ServerCertificateNotYetValid,
        ClientCertificateRejected,
        ClientCertificateRequired,
        CannotLoadFromNetwork,
        CannotCreateFile,
        CannotOpenFile,
        CannotCloseFile,
        CannotWriteToFile,
        CannotRemoveFile,
        CannotMoveFile,
        DownloadDecodingFailedMidStream,
        DownloadDecodingFailedToComplete,
        InternationalRoamingOff,
        CallIsActive,
        DataNotAllowed,
        RequestBodyStreamExhausted,
        BackgroundSessionRequiresSharedContainer,
        BackgroundSessionInUseByAnotherProcess,
        BackgroundSessionWasDisconnected,
    }
    #[allow(unused_imports)]
    use super::*;
    #[cfg(feature = "Foundation")]
    #[allow(unused_imports)]
    use icrate::Foundation::*;
    #[allow(unused_imports)]
    use objc2::{rc::Id, runtime::Object};
    #[derive(Default)]
    pub struct UserInfo {
        pub failing_url: Option<Id<NSURL>>,
        pub failing_url_string: Option<Id<NSString>>,
        pub background_task_cancelled_reason:
            Option<ns_url_error_reasons::BackgroundTaskCancelledReason>,
        pub network_unavailable_reason: Option<ns_url_error_reasons::NetworkUnavailableReason>,
    }
}
```

**Comment by madsmtm** [comment]
I think it's important to separate **syntax** of the macro from the **desired output**. To this end, I've edited the top-level comment with what I think the desired output should be, then that part is perhaps easier to discuss.
For example, I disagree that we want `NSNumber` conversions, I think it's better to just have raw primitive conversions, and then the user can do `number.as_isize()` themselves if they need to.

---

I can see you've spent a lot of time on translation of error enums - I'm quite uncertain about those myself, but I think you've taken a big step in the right direction. Comments on that:
- `Cases` feels a bit redundant, couldn't we have just `Codes`? The user can't do an exhaustive match anyway.
- I like the idea of allowing easier access to user info keys, but I think it should rather be a zero-cost solution like this (e.g. instead of fetching everything into a `struct`, we allow the user to only fetch the keys that are relevant to them):
    ```rust
    impl NSURLError {
        pub fn userInfo() -> UserInfo { ... }
    }
    
    struct UserInfo(NSDictionary<NSErrorUserInfoKey, Object>);
    // impl RefEncode + Message + Deref + ...
    
    impl UserInfo {
        pub fn failing_url() -> Option<Id<NSURL>> { ... }
        pub fn failing_url_string() -> Option<Id<NSString>> { ... }
        // Or perhaps just returns `Option<Id<NSNumber>>`
        pub fn background_task_cancelled_reason() -> Option<BackgroundTaskCancelledReason> { ... }
        // Or perhaps just returns `Option<Id<NSNumber>>`
        pub fn network_unavailable_reason() -> Option<Id<NetworkUnavailableReason>> { ... }
    }
    ```

    Though I recognize that that doesn't lend itself as easily to the `new_with_user_info` pattern :/
- I we should consider getting rid of the module, and just have `NSURLError`, `NSURLErrorCode` and `NSURLErrorUserInfo`.

---

Regarding macro syntax, I think it's basically irrelevant since I don't think it'll really be necessary to expose the error enum macro to users? In any case, I think I'd like to split the error and the `userInfo` stuff into two different macros, it seems like they don't need to be in the same macro?

**Comment by silvanshade** [comment]
> For example, I disagree that we want `NSNumber` conversions, I think it's better to just have raw primitive conversions, and then the user can do `number.as_isize()` themselves if they need to.

Without the `NSNumber` conversions, we cannot automatically generate the `UserInfo` machinery without requiring that the user write `from` and `into` for every instance where the field is an some generated enum type.

That's the only reason they are there really.

We could easily make them optional as well, for cases where they wouldn't be needed.

> I can see you've spent a lot of time on translation of error enums - I'm quite uncertain about those myself, but I think you've taken a big step in the right direction.

I mostly followed the design and rationale that was laid out [here](https://github.com/apple/swift-evolution/blob/main/proposals/0112-nserror-bridging.md).

Personally, I think we really should have *something* like this if the intention is truly to allow Rust (through this crate ecosystem) to be an alternative to Swift in terms of ergonomics and feature parity (as far as that is possible).

Did you have another alternative in mind perhaps?

> * `Cases` feels a bit redundant, couldn't we have just `Codes`? The user can't do an exhaustive match anyway.

The main motivation for `Cases` is to provide a better pattern matching experience through various Rust tooling. In particular, I heavily use the Rust Analyzer feature for automatically filling in match arms, and this feature wouldn't work if we only had `Codes`.

The experience of just pressing a key combination and getting all of the error cases filled in versus looking them up individually in the docs (or even just completing each one manually) makes a fairly big difference in usability from my perspective.

> * I like the idea of allowing easier access to user info keys, but I think it should rather be a zero-cost solution like this (e.g. instead of fetching everything into a `struct`, we allow the user to only fetch the keys that are relevant to them):

We have both actually. I generate individual getters first, then the code that fetches the entire dictionary is just implemented in terms of that:

```rust
pub fn user_info(&self) -> ns_url_error::UserInfo {
    let mut user_info = ns_url_error::UserInfo::default();
    user_info.failing_url = self.failing_url();
    user_info.failing_url_string = self.failing_url_string();
    user_info.background_task_cancelled_reason = self.background_task_cancelled_reason();
    user_info.network_unavailable_reason = self.network_unavailable_reason();
    user_info
}
```

> * I we should consider getting rid of the module, and just have `NSURLError`, `NSURLErrorCode` and `NSURLErrorUserInfo`.

I think that would probably be fine. Part of the rationale behind the use of the submodules was to allow for fewer imports (by relying on the `Cases`, `Codes`, and `UserInfo` helper types being in scope already). I'm not sure how much of a difference that will make on average though.

> Regarding macro syntax, I think it's basically irrelevant since I don't think it'll really be necessary to expose the error enum macro to users?

I'm not sure actually. Do people use `ns_error_enum` in 3rd party Objective-C frameworks? I guess the other question is whether we intend to support that use case in general. I've always kind of assumed we would but I don't know how the translator would work for other frameworks.

> In any case, I think I'd like to split the error and the `userInfo` stuff into two different macros, it seems like they don't need to be in the same macro?

It would be possible to split the `userInfo` support code generation into a separate macro invocation.

The way the code is implemented, they are not deeply intertwined. Mostly the `ns_error_enum` macro is a combination of `ns_typed_extensible_enum` and `userInfo` generation. I just felt at the time that it was the most logical place for that information (following the general design of other attribute proc-macros). But maybe there is a better alternative.

**Comment by silvanshade** [comment]
> I think it's important to separate **syntax** of the macro from the **desired output**. To this end, I've edited the top-level comment with what I think the desired output should be, then that part is perhaps easier to discuss.

Thanks for the examples. That gives a good point of reference. I'll comment on the individual samples versus what is currently implemented.

- `NS_OPTIONS`

I am currently using the `bitflags!` macro like you suggest. I haven't the `objc2` helper traits yet.

With regard to whether we want all the additional helper traits from `bitflags`, I would personally opt to use what they provide unless we find that it impacts compilation time too severely.

I think the maintenance burden of having to maintain our own reimplementation is probably not worth it. But that's just my personal take.

- `NS_ENUM`

I am currently translating this as an enum, but was worried about the exact issue you mention (which I hadn't tested yet).

I think the best option is just to handle this case the same as `NS_TYPED_EXTENSIBLE_ENUM`, specialized for numeric types. This is basically what I am doing for the `NS_ERROR_ENUM` case.

- `NS_CLOSED_ENUM`

I am currently translating this how you suggest but haven't added the `objc2` helper traits yet.

- `NS_TYPED_ENUM`, `NS_TYPED_EXTENSIBLE_ENUM`

The representation you propose for both of these is different to how I am currently generating them.

For starters I am using `#[repr(transparent)]` instead of `#[repr(C)]`, with the intention to facilitate transmutes. Is there a need for `#[repr(C)]` instead for some reason?

The other part that is different is that I am (or was) using `const` instead of `fn` for defining constants. But I realized from your example that that is problematic in the case where the variants refer to statics (since constants can't refer to statics).

So I can change that part. But that also brings me back to the issue with pattern matching, and why I think helper `Cases` enum is important for ergonomics, especially even more so because we can't use `const` here anymore, so we can't even do partial matches.

Generating `Cases` is more complicated in for the `NS_TYPED_EXTENSIBLE_ENUM` scenario but there are ways around it. This would be another instance where `linkme` would be useful, but we could also handle the situation manually, or with some sort of collation done during the header translation.

- `NS_ERROR_ENUM` with typedef vs. without typedef

I'm not sure there is an important distinction between these cases, other than being an artifact of the conventions used for writing the headers at various points in time.

Sometimes the headers are not using the most appropriate definitions or contain other issues. For instance, sometimes they are not using the `NSErrorDomain` type (see [here](https://developer.apple.com/documentation/soundanalysis/snerrordomain?language=objc)) for seemingly no particular reason. (In this case, it causes us problems because we translate [the result](https://docs.rs/icrate/latest/icrate/SoundAnalysis/static.SNErrorDomain.html) with an option).

I think we should be looking at what the Swift analogue is for such cases and use that as a point of reference for our desired result.

In that case, [MTLIOError](https://developer.apple.com/documentation/metal/mtlioerror) and [URLError](https://developer.apple.com/documentation/metal/mtlioerror) are defined similarly.

**Comment by silvanshade** [comment]
I'll work on adding some of the stuff currently missing that you refer to in the original comment and then add some more examples based on that and see how it turns out.

**Comment by madsmtm** [comment]
> > I can see you've spent a lot of time on translation of error enums - I'm quite uncertain about those myself, but I think you've taken a big step in the right direction.
> 
> I mostly followed the design and rationale that was laid out [here](https://github.com/apple/swift-evolution/blob/main/proposals/0112-nserror-bridging.md).
> 
> Personally, I think we really should have _something_ like this if the intention is truly to allow Rust (through this crate ecosystem) to be an alternative to Swift in terms of ergonomics and feature parity (as far as that is possible).
> 
> Did you have another alternative in mind perhaps?

Hmm, perhaps I sounded dismissive there? I agree that we should have something like this!

> The main motivation for `Cases` is to provide a better pattern matching experience through various Rust tooling.

Hmm, I never considered that, I agree that would be quite nice to have! My problems with it is:
- I think we should provide one "recommended" way of dealing with enums; e.g. I think it's overkill and confusing for the user if we have two types pr. enum. (Also the reason I'm using `objc2::rc::Id` and `objc2::rc::Allocated` instead of e.g. `NSObjectRef<'_>`, `NSObjectOwned` and `NSObjectAllocated`).
- I think people will want to use `Cases`, which in turn will lead to a lot of `unwrap`-ing of the `Option`.
- Using `Cases` as it is currently is fundamentally a pit-fall, since it doesn't allow a round-trip - e.g. if you wanted to create a new `NSEvent` from an old one, using `NSEventTypeCases` would be wrong.

I think it would be possible to do something like this:
```rust
#[non_exhaustive]
enum NSFoo {
    A = 1,
    B = 2,
    #[doc(hidden)]
    __Unknown(usize),
}

impl EncodeConvertArgument for NSFoo { ... }
impl EncodeConvertResult for NSFoo { ... }
```

Which would allow taking and returning `NSFoo` from functions (but not in other places like blocks). But then again, that increases the size of the enum, making it no longer zero-cost!
What we really need is for Rust to provide [a way to specify niches](https://github.com/rust-lang/rfcs/pull/3334) for the unknown variant.

> > * I like the idea of allowing easier access to user info keys, but I think it should rather be a zero-cost solution like this (e.g. instead of fetching everything into a `struct`, we allow the user to only fetch the keys that are relevant to them):
> 
> We have both actually. I generate individual getters first, then the code that fetches the entire dictionary is just implemented in terms of that:

Yeah, but that calls the `userInfo` selector multiple times (e.g. first inside `failing_url`, next inside `failing_url_string`, ...).

> I don't know how the translator would work for other frameworks.

I haven't planned for `header-translator` to ever be public, since I just haven't really seen a use-case for 3rd party Objective-C frameworks? Like, the vast majority of Objective-C code is Apple's, and I suspect the remaining frameworks implement something that is already done better by other Rust crates?
Though I did just now find [awesome iOS](https://github.com/vsouza/awesome-ios), so perhaps there _are_ things there that people will want to use? Are there libraries/frameworks that _you_ want to use?

> It would be possible to split the `userInfo` support code generation into a separate macro invocation.

I think they should be separate, mostly because it seems like the error stuff we can autogenerate, while the `userInfo` stuff we can't?

> I am using `#[repr(transparent)]` instead of `#[repr(C)]`, with the intention to facilitate transmutes. Is there a need for `#[repr(C)]` instead for some reason?

No, they act exactly the same for structs with one field - the reason I wrote `#[repr(C)]` was because [in the past you couldn't use `#[repr(transparent]` on zero-sized types](https://rust.godbolt.org/z/cM69nzTK9) (and [the reference still documents this as well](https://doc.rust-lang.org/reference/type-layout.html#the-transparent-representation)).

> `NS_TYPED_ENUM` and the issue with pattern matching

Hmm, yeah, performance is a bit less of a concern here, since we're dealing with string comparisons which are comparatively slow anyhow. What you've done seems fairly optimal, though I still dislike that we're emitting two types :/

> `NS_ERROR_ENUM` with typedef vs. without typedef

I think I was mostly thinking of it in terms of what we can autogenerate, and what can't - e.g. `URLError` is created using a [manual Swift wrapper](https://github.com/apple/swift-corelibs-foundation/blob/f3f3a73ba9a4d71da6368ddec319ba77ec912eea/Sources/Foundation/NSError.swift#L850).

> `AsRef`, `Borrow`, and similar for `NS_TYPED_ENUM`

I think it's worth it to special-case `NSString` for these, since it's so common.

**Comment by silvanshade** [comment]
> Hmm, perhaps I sounded dismissive there? I agree that we should have something like this!

Great! And by the way, I do agree there are a number of issues with the current implementation. I'm just hoping that there's a good idea in here somewhere that we can figure out eventually.

> Yeah, but that calls the `userInfo` selector multiple times (e.g. first inside `failing_url`, next inside `failing_url_string`, ...).

Okay, I understand what you mean better now.

We could instead specialize the `user_info` getter during code-generation to avoid multiple calls to the underlying selector and then advise users that if they intend to use multiple fields it would possibly be more efficient to get the whole `UserInfo` that way versus using individual getters.

That seems like a fairly reasonable compromise I think? (versus tending toward a more complex design)

> * I think we should provide one "recommended" way of dealing with enums; e.g. I think it's overkill and confusing for the user if we have two types pr. enum. (Also the reason I'm using `objc2::rc::Id` and `objc2::rc::Allocated` instead of e.g. `NSObjectRef<'_>`, `NSObjectOwned` and `NSObjectAllocated`).

That is a fair point and I understand the comparison. I'm not entirely sure I agree that it's confusing to the same degree though. 

Memory management and FFI-related code are difficult due to the complex and very situational semantics involved. The cognitive surface area with regard to API design is also generally much larger.

In comparison, the user doesn't have to understand too much about the struct-wrapping-values and enum-representing-match-cases pairing. The interface is limited: there's only a single method (`.cases()`) and you really only use it in a single situation, when pattern matching.

> * Using `Cases` as it is currently is fundamentally a pit-fall, since it doesn't allow a round-trip - e.g. if you wanted to create a new `NSEvent` from an old one, using `NSEventTypeCases` would be wrong.

I'm not sure I would expect the user to try to use `Cases` this way.

Consider the following code snippet:

```rust
fn generate_event(event: &NSEvent) -> Option<Id<NSEvent, Shared>> {
    match event.type().cases() {
        Some(NSEventTypeCases::LeftMouseDown) => {
            NSEvent::otherEventWithType_location_modifierFlags_timestamp_windowNumber_context_subtype_data1_data2(
                type: NSEventType::application_defined(),
                ...
            )
        }
        Some(NSEventTypeCases::KeyDown) => {
            ...
        }
        _ => {
            None
        }
    }
}
```

I suppose there might be a situation where you really want to generate the `NSEventType` value from the `NSEventTypeCases`, and it's possible to define a `impl From<NSEventTypeCases> for NSEventType` for that scenario. I originally included that in the output.

But the more I thought about it, the less likely it seemed that someone would really want to use that in practice, since any time you get the `Cases` (by calling `.cases()`), you should already have a value for the respective newtype wrapper at hand.

Maybe there is an example of that you had in mind?

> * I think people will want to use `Cases`, which in turn will lead to a lot of `unwrap`-ing of the `Option`.

Yeah, this one is something I've had concerns about also.

I did consider an alternate design that returned a `Result` instead of an `Option`, so that the user could more often use `?` to unwrap (or use `.cases().ok()` if they want the `Option`), but that complicates the design. But maybe it's worth considering that more seriously?

On the other hand, the user doesn't *have* to unwrap, in the sense of explicitly writing a separate branch.

It's still possible to write this, which isn't much more verbose:

```rust
match value.cases() {
    Some(EnumCases::Var0) => ...,
    Some(EnumCases::Var1) => ...,
    _ => ...,
}
```

In comparison, this is what it would look like *without* `.cases()`:

```rust
match value {
    _ if value == Enum::var0() => ...,
    _ if value == Enum::var1() => ...,
    _ => ...,
}
```

This seems more verbose and you also lose the exhaustivity checking and tooling support for filling in the cases.

However, despite advocating for `Cases`, I'm also not entirely convinced that it's a good idea to include. I just find it slightly bothersome to be able to get part of the way toward replicating what Swift is able to do for these situations, but missing the pattern matching, since pattern matching is a big part of what makes Rust so nice to use.

For the sake of moving this forward, I could disable (or make opt-in), the `Cases` generation functionality.

> Like, the vast majority of Objective-C code is Apple's, and I suspect the remaining frameworks implement something that is already done better by other Rust crates?

That's a good point and I think you are probably right in general.

The one case that I had in mind where it might be relevant beyond this project is for commercial users of the crate. I could certainly imagine a developer with a large amount of code written in Objective-C that might be able to make use of `header-translator` for their company-specific proprietary libraries.

> Are there libraries/frameworks that _you_ want to use?

Not specifically. There are some interesting frameworks like [ResearchKit](https://github.com/ResearchKit/ResearchKit) that would be nice to translate but nothing I would probably use for my own projects at this point.

> No, they act exactly the same for structs with one field - the reason I wrote `#[repr(C)]` was because [in the past you couldn't use `#[repr(transparent]` on zero-sized types](https://rust.godbolt.org/z/cM69nzTK9) (and [the reference still documents this as well](https://doc.rust-lang.org/reference/type-layout.html#the-transparent-representation)).

Ah, okay, good to know.

> I think I was mostly thinking of it in terms of what we can autogenerate, and what can't - e.g. `URLError` is created using a [manual Swift wrapper](https://github.com/apple/swift-corelibs-foundation/blob/f3f3a73ba9a4d71da6368ddec319ba77ec912eea/Sources/Foundation/NSError.swift#L850).

Makes sense. I guess I was imagining that we would probably need to do this already for a lot of the error enums in order to have nice interfaces. But it is important to consider how much of it we could autogenerate for sure.

Relatedly:

> I think they should be separate, mostly because it seems like the error stuff we can autogenerate, while the `userInfo` stuff we can't?

Okay, that makes sense also. I already started working on providing an alternative interface where that macro is separate so it will probably be in the PR branch soon.

> I think it's worth it to special-case `NSString` for these, since it's so common.

Okay, that's doable.

Anyway, thanks for the feedback! You definitely raised a number of good points.

**Comment by madsmtm** [comment]
Update on this:
- Enum constants are now on the enum itself, see https://github.com/madsmtm/objc2/pull/580.
- `NS_OPTIONS` with `bitflags!` support was tracked in https://github.com/madsmtm/objc2/issues/602 and implemented in https://github.com/madsmtm/objc2/commit/e87335b50efdb5a2ed45b2db2c680575bae267c5.
- I decided against conversion methods on `NS_ENUM`, and just exposed the inner field directly - they're safe to construct from a raw value in Swift.
- `NS_TYPED_ENUM`, `NS_TYPED_EXTENSIBLE_ENUM` and `NS_ERROR_ENUM` are still fairly unsupported.

This means that nowadays, the "`generate_event`" example you gave would look like this:

```
fn generate_event(event: &NSEvent) -> Option<Id<NSEvent, Shared>> {
    match event.type() {
        NSEventType::LeftMouseDown => {
            NSEvent::otherEventWithType_location_modifierFlags_timestamp_windowNumber_context_subtype_data1_data2(
                type: NSEventType::application_defined(),
                ...
            )
        }
        NSEventType::KeyDown => {
            ...
        }
        _ => {
            None
        }
    }
}
```

Your IDE would not autofill the cases, doing so is still blocked on better niche support in Rust.

**Comment by madsmtm** [comment]
Linking a bit of discussion on non-exhaustive C-like enums on Zulip recently, see [this topic](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Design.20for.20an.20.22enum.20with.20unknown.20variants.22.20feature).
